### PR TITLE
type-checking investigation

### PR DESCRIPTION
**Pull request by mathematicalmichael**
https://github.com/mathematicalmichael/mud/actions/runs/3425141460/jobs/5705619794 failed in a way I can't seem to reproduce locally.

`mypy src/mud tests` works fine for me, but perhaps there's a version difference in the github action that can explain things.

In the runner:
```
Python 3.10.8
mypy 0.990
...
Found 32 errors in 7 files (checked 24 source files)
```

In my env:
```
Python 3.9.7
mypy 0.982
...
Found 0 errors
```

upgrading to `mypy==0.990` reproduces the error. This is a recent release.
- of the 32 errors, 10 of them are implicit optional for `save_path`
  - currently `save_figure` acts more as "maybe save figure" and that's confusing, so while it's inefficient to do so (rather than renaming the function to include an awkward prepended `maybe_`), I'm updating the logic for whether to call this function to the calling function. i.e. I'm adding a bunch of `if save_path:` indent blocks, this avoids changing the defaults /docstrings and allows for the option to use the type hint `Optional[str]`.
  - one function had two unused arguments, so I removed them
  - refactored functions to reduce "complexity" according to flake8.
- of the remaining 22 errors... all fixed automatically with the `no_implicit_optional` tool.

### CODE REVIEW

File: src/mud/examples/adcirc.py [hunk 1]
```diff
@@ -112,10 +112,10 @@ def tri_mesh_plot(
                 label=f"Recording Station {i}",
             )
     ax.legend()
-
-    save_figure(
-        f"si-{value}", save_path, close_fig=close_fig, dpi=dpi, bbox_inches="tight"
-    )
+    if save_path:
```
Comment (cdelcastillo21) [inline src/mud/examples/adcirc.py:115]: Should be 
`if save_path is not None:`
See [PEP 8](https://peps.python.org/pep-0008/)

File: src/mud/examples/adcirc.py [hunk 2]
```diff
@@ -171,22 +171,26 @@ def adcirc_ts_plot(
     ax.set_ylabel("Water Elevation (m)")
     ax.set_xlabel("Time")
     _ = ax.set_title("")
-
-    save_figure(
-        "adcirc_full_ts", save_path, close_fig=close_fig, dpi=dpi, bbox_inches="tight"
-    )
+    if save_path:
```
Comment (cdelcastillo21) [inline src/mud/examples/adcirc.py:174]: Should be 
`if save_path is not None:`
See [PEP 8](https://peps.python.org/pep-0008/)

File: src/mud/examples/adcirc.py [hunk 3]
```diff
@@ -218,22 +222,24 @@ def adcirc_time_window(
...
             adcirc_prob, t_mask, msize=msize, max_plot=max_plot, title=title
         )
-        save_figure(
-            f"pca_vecs_{num_components}_{ndata}",
-            save_path,
-            close_fig=close_fig,
-            dpi=dpi,
-            bbox_inches="tight",
-        )
+        if save_path:
```
Comment (cdelcastillo21) [inline src/mud/examples/adcirc.py:225]: Should be 
`if save_path is not None:`
See [PEP 8](https://peps.python.org/pep-0008/)

File: src/mud/examples/linear.py [hunk 4]
```diff
@@ -676,12 +703,13 @@ def run_wme_covariance(
     _ = ax.set_xlabel("Index")
     _ = ax.set_ylabel("Eigenvalue")
     _ = ax.legend(loc="lower left")
-    save_figure(
-        "lin-meas-cov-sd-convergence.png",
-        save_path=save_path,
-        dpi=dpi,
-        close_fig=close_fig,
-    )
+    if save_path:
```
Comment (cdelcastillo21) [inline src/mud/examples/linear.py:706]: Should be 
`if save_path is not None:`
See [PEP 8](https://peps.python.org/pep-0008/)

File: src/mud/examples/linear.py [hunk 5]
```diff
@@ -800,13 +828,13 @@ def err(xs):
...
     _ = ax.set_xlabel("Dimension of Output Space")
     _ = ax.legend(["MUD", "MAP", "Least Squares"])
-
-    save_figure(
-        "lin-dim-cov-convergence.png",
-        save_path=save_path,
-        dpi=dpi,
-        close_fig=close_fig,
-    )
+    if save_path:
```
Comment (cdelcastillo21) [inline src/mud/examples/linear.py:831]: Should be 
`if save_path is not None:`
See [PEP 8](https://peps.python.org/pep-0008/)

File: src/mud/examples/linear.py [hunk 6]
```diff
@@ -840,12 +868,12 @@ def err(xs):
...
     _ = ax.set_xlabel("Rank(A)")
     _ = ax.legend(["MUD", "MAP", "Least Squares"])
-
-    save_figure(
-        "lin-rank-cov-convergence.png",
-        save_path=save_path,
-        dpi=dpi,
-        close_fig=close_fig,
-    )
+    if save_path:
```
Comment (cdelcastillo21) [inline src/mud/examples/linear.py:871]: Should be 
`if save_path is not None:`
See [PEP 8](https://peps.python.org/pep-0008/)

File: src/mud/examples/poisson.py [hunk 7]
```diff
@@ -126,13 +126,99 @@ def spline_objective_function_2d(lam, aff=10000):
...
 
 
+def plot_qoi_ex(mud_prob, num_components: int, **kwargs):
+    fig = plt.figure(figsize=(10, 5))
+    for i in range(num_components):
+        ax = fig.add_subplot(1, 2, i + 1)
+        mud_prob.plot_params_2d(ax=ax, y=i, contours=True, colorbar=True)
+        if i == 1:
+            ax.set_ylabel("")
+    if kwargs.get("save_path"):
```
Comment (cdelcastillo21) [inline src/mud/examples/poisson.py:136]: Should be 
`if kwargs,get("save_path") is not None:`
See [PEP 8](https://peps.python.org/pep-0008/)

File: src/mud/examples/poisson.py [hunk 8]
```diff
@@ -126,13 +126,99 @@ def spline_objective_function_2d(lam, aff=10000):
...
+    plot_vert_line(ax, 2 / 3.0, color="b", linestyle="--", alpha=0.6)
+
+    ax.set_title("Boundary Condition")
+    ax.set_ylabel("")
+    ax.legend(
+        ["$g(x_2)$", r"$\hat{g}(x_2,\lambda^\dagger)$", r"$\hat{g}(x_2,\lambda_i)$"]
+    )
+
+    fig.tight_layout()
+    if kwargs.get("save_path"):
```
Comment (cdelcastillo21) [inline src/mud/examples/poisson.py:210]: Should be 
`if kwargs,get("save_path") is not None:`
See [PEP 8](https://peps.python.org/pep-0008/)

File: src/mud/examples/poisson.py [hunk 9]
```diff
@@ -196,95 +282,42 @@ def run_2d_poisson_sol(
...
+            dpi=dpi,
+            bbox_inches="tight",
         )
         axes.append(ax)
     if "densities" in plot_fig or "all" in plot_fig:
         ax1 = mud_prob.plot_param_space(param_idx=0, **param1_kwargs)
-        save_figure(
-            "lam1", save_path, close_fig=close_fig, dpi=dpi, bbox_inches="tight"
-        )
+        if save_path:
```
Comment (cdelcastillo21) [inline src/mud/examples/poisson.py:311]: Should be 
`if save_path is not None:`
See [PEP 8](https://peps.python.org/pep-0008/)

File: src/mud/examples/poisson.py [hunk 10]
```diff
@@ -196,95 +282,42 @@ def run_2d_poisson_sol(
...
+            save_figure(
+                "lam1", save_path, close_fig=close_fig, dpi=dpi, bbox_inches="tight"
+            )
 
         ax2 = mud_prob.plot_param_space(param_idx=1, **param2_kwargs)
-        save_figure(
-            "lam2", save_path, close_fig=close_fig, dpi=dpi, bbox_inches="tight"
-        )
+        if save_path:
```
Comment (cdelcastillo21) [inline src/mud/examples/poisson.py:317]: Should be 
`if save_path is not None:`
See [PEP 8](https://peps.python.org/pep-0008/)

File: src/mud/examples/poisson.py [hunk 11]
```diff
@@ -419,14 +452,14 @@ def run_2d_poisson_trials(
...
         ax2.legend()
-
-        save_figure(
-            f"solution_n{N}",
-            save_path,
-            close_fig=close_fig,
-            dpi=dpi,
-            bbox_inches="tight",
-        )
+        if save_path:
```
Comment (cdelcastillo21) [inline src/mud/examples/poisson.py:455]: Should be 
`if save_path is not None:`
See [PEP 8](https://peps.python.org/pep-0008/)

**Review by cdelcastillo21 — CHANGES_REQUESTED** [review]
Just `is not None:` checks according to [PEP 8](https://peps.python.org/pep-0008/). I think I marked them all.

File: src/mud/examples/adcirc.py (hunk 1 above)
Comment (mathematicalmichael) [reply src/mud/examples/adcirc.py:115]: hm, why doesn't flake8 complain about this?
my thinking was actually to check for both `None` and `""` simultaneously.
Though perhaps the behavior of `""` is fine... realized I'm not actually sure what would happen (would it try to save to root or to cwd?)

Comment (mathematicalmichael) [reply src/mud/examples/adcirc.py:115]: > Also, beware of writing if x when you really mean if x is not None – e.g. when testing whether a variable or argument that defaults to None was set to some other value. The other value might have a type (such as a container) that could be false in a boolean context!

the type is `str`, and I did intend for `""` to evaluate in the same context (this is the only `str` that evals to False), so I was "careful" in the sense that this recommendation suggests. Question is... was I right about my assumption that `""` is a bad argument to pass to `save_figure`? I would prefer "cwd" to be given by `"."` which evals to True.

Comment (cdelcastillo21) [reply src/mud/examples/adcirc.py:115]: Either works. I think PEP8 is just there to make the code more clear and readable in this case. "" evaluates to None but if not explicitly stated it just obfuscates the code a bit.

Comment (mathematicalmichael) [reply src/mud/examples/adcirc.py]: fwiw, i ended up doing this.

**Pull request merged**